### DISCUSSION TITLE
Multi-drone/Swarm using Motion Commander

### DISCUSSION THREAD

**Discussion by john-amadora** [bitcraze/discussions#1571] [Q&A]
Hi! 

I am working on a control interface for multiple drones at a time. Not specifically swarming. So I wanted to know how I could send motioncommander commands to multiple (at most 3) drones. So the following cases can happen

1. Each drone can be given a different command (ex: drone 1 goes left while drone 2 goes right)
2. All drones can be given a motion commander command to execute together.

**Comment by ataffanel** [comment]
Hi,

Both case are possible in Python using the Crazyflie python lib, however I do not think the ```Swarm``` helper of the lib is the right tool for you.

The ```Swarm``` helper has been designed to drive a swarm of Crazyflie in simple cases where all the Crazyflies will do the same thing. In your case it would be better to implement the Crazyflie handling by yourself.

The exact implementation depends of lot of your usecase, you can spawn one python thread per Crazyflie and talk to them with command queues or keep all Crazyflies in an array and have a main-loop sending commands to them.

On the lib side, each connection to a Crazyflie is a python object that you will have to handle and keep alive (ie. dropping the Crazyflie object out of memory will close the connection).

**Reply by john-amadora** [reply]
HI @ataffanel , this works! But I am not satisfied with the delay between the drones (i am currently testing with two) 

So I tried multiprocessing and I keep getting errors. Would you happen to know if multiprocessing works with handling multiple  synccrazyflies using motioncommander?

**Reply by ataffanel** [reply]
How much delay between drones are you observing with 2 connections? There should not be more than a couple of 10th of milliseconds delay when talking to 2 drones.

**Reply by john-amadora** [reply]
It's about 1 or 2 seconds. Here's my code. Its a very simple code in order to test threading. 

I used the time.sleep within the main thread in order to make sure both drones are already okay with the SynCrazyFlie

Then both threads just use the same function but with different drones.
The latency between the drones is about 1 or 2 seconds. 
```
def flydrone(scf, drone):
    with MotionCommander(scf, default_height=DEFAULT_HEIGHT) as mc:
            print(f"{drone}: Going up!")
            mc.up(0.5)
            print(f"{drone}: Going left!")
            mc.left(0.5)
            print(f"{drone}: Going right!")
            mc.right(0.5)
            print(f"{drone}: Sleeping!")
            time.sleep(5)
            print(f"{drone}: Stopping!")
            mc.stop()

if __name__ == '__main__':
    cflib.crtp.init_drivers()

    with SyncCrazyflie(URI1, cf=Crazyflie(rw_cache='./cache')) as scf1, SyncCrazyflie(URI2, cf=Crazyflie(rw_cache='./cache')) as scf2:
        time.sleep(5)
            
        thread1 = threading.Thread(target=flydrone, args=(scf1, "Drone 1",))
        thread2 = threading.Thread(target=flydrone, args=(scf2, "Drone 2",))
        thread1.start()
        thread2.start()
        thread1.join()
        thread2.join()
        print("Simulation Done")

```

I have tried multiprocessing instead but all I get is errors, has this been tested?

**Reply by ataffanel** [reply]
I think you need to use synchronization routines if you want both threads to be synchronized, for example you could use a [barrier](https://docs.python.org/3/library/threading.html#barrier-objects) to lock-step each thread at the same point calling .wait() after each movement. I assume this will make the control much more synchronized.
